### PR TITLE
first stab at cells dict -> list of (named) tuples

### LINKED ISSUE #629: cells dict -> list of tuple/namedtuple/dataclasses?
Another idea for 4.0:

We could transform the cells (and cell data) from a dict to a list of tuples, namedtuples, or dataclasses ([as has been suggested before](https://github.com/nschloe/meshio/issues/615#issuecomment-570833763)), having the attributes `type` (e.g., `triangle6`) and `data` (a numpy array). With this, we could preserve cell ordering (@MuellerSeb is going to like that) and better represent data structures like Exodus's regions.

Opinions, please. @tianyikillua

### PR DESCRIPTION
**Pull request by nschloe**
Fixes #629.

### CODE REVIEW

**Comment by nschloe** [comment]
Side note: It turns out h5py doesn't preserve the group ordering yet, cf. https://github.com/h5py/h5py/issues/1471, so a write-read might produce a different cell ordering.

**Comment by tianyikillua** [comment]
For the MED format, I don't think it is possible to do
```
for cell_type, cells in mesh.cells:
    med_type = meshio_to_med_type[cell_type]
    med_cells = cells_group.create_group(med_type)
```
since the MED format doesn't support multiple same cell types. We have to convert `mesh.cells` to the former `dict` structure, that is to aggregate all cells with the same type.

**Comment by nschloe** [comment]
This is basically done, I think I'll merge after dinner. I'm quite happy with the PR, especially since the new approach allowed to remove some costly `unique` calls on the cell types for VTK, VTU, XDMF.

File: meshio/_mesh.py [hunk 1]
```diff
@@ -114,6 +114,13 @@ def write(self, path_or_buf, file_format=None, **kwargs):
 
         write(path_or_buf, self, file_format, **kwargs)
 
+    @property
+    def cells_dict(self):
+        assert len(self.cells) == len(
+            numpy.unique([c.type for c in self.cells])
+        ), "More than one block of the same type. Cannot create dictionary."
+        return {cell_type: data for cell_type, data in self.cells}
```
Comment (gdmcbain) [inline meshio/_mesh.py]: An equivalent:
```python
return dict(self.cells)
```

Comment (nschloe) [reply meshio/_mesh.py]: Much better, thanks.

**Comment by tianyikillua** [comment]
Great work!

I suppose this PR should be accompanied by a another that updates `cell_data["name"]` also to a list of the same length as `cells`, that contains numpy cell values?

**Comment by nschloe** [comment]
Cell data is also changed in this one, exactly to match `cells`. Now, `cell_data["name"]` is a list of arrays of the same length as `mesh.cells`, and `len(cell_data["name"][k]) == len(cells[k].data)`.

File: meshio/gmsh/_gmsh41.py [hunk 2]
```diff
@@ -216,31 +216,15 @@ def _read_elements(f, point_tags, physical_tags, is_ascii, data_size):
...
-                    cell_data["gmsh:physical"][key] = {}
-                cell_data["gmsh:physical"][key] = physical_tag[0] * numpy.ones(
-                    len(values), int
-                )
+        cells.append((key, values))
+        if physical_tag:
+            if "gmsh:physical" not in cell_data:
+                cell_data["gmsh:physical"] = {}
+            cell_data["gmsh:physical"].append(
+                (key, physical_tag[0] * numpy.ones(len(values), int))
```
Comment (dalcinl) [inline meshio/gmsh/_gmsh41.py]: @nschloe Why you do `(key, tags)` in here? Using `tags` would be just fine, the cell type in here is redundant, and hardly useful if you have many blocks with the same cell type. The cell data ordering is implicit from the `mesh.cells` attribute. This broke compatibility with the gmsh 2.2 reader/writer, particularly if you want to read v4.1 and dump v2.2 (which I was using routinely before these changes). I'm working on fixing this issue. 

Would you accept this change (plus other stuff to make conversion from 4.1 to 2.2)?

```diff
--- a/meshio/gmsh/_gmsh41.py
+++ b/meshio/gmsh/_gmsh41.py
@@ -222,7 +222,7 @@ def _read_elements(f, point_tags, physical_tags, is_ascii, data_size):
             if "gmsh:physical" not in cell_data:
                 cell_data["gmsh:physical"] = []
             cell_data["gmsh:physical"].append(
-                (key, physical_tag[0] * numpy.ones(len(values), int))
+                physical_tag[0] * numpy.ones(len(values), int)
             )
```

Comment (nschloe) [reply meshio/gmsh/_gmsh41.py]: >  Using tags would be just fine

Yep, that's an oversight. PR welcome!

**Pull request merged**